homework5/MyNUnit/MyNUnit/MyNUnit.cs

```diff
@@ -0,0 +1,178 @@
+﻿namespace MyNUnitSpace;
+
+using Attributes;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Reflection;
+
+/// <summary>
+/// NUnit class
+/// </summary>
+public class MyNUnit
+{ 
+    /// <summary>
+    /// State of test
+    /// </summary>
+    private enum TestState
+    {
+        Ignored,
+        Errored,
+        Failed,
+        Passed
+    };
+
+    /// <summary>
+    /// Concurrent queue that stores information about tests
+    /// </summary>
+    public readonly ConcurrentQueue<TestInfo> testsInfo;
+
+    public MyNUnit(string path)
+    {
+        var files = Directory.GetFiles(path, "*.dll", SearchOption.AllDirectories);
+        var filesWithoutRepeats = new Dictionary<string, string>();
+        foreach (var file in files)
+        {
+            if (!filesWithoutRepeats.ContainsValue(Path.GetFileName(file)))
+            {
+                filesWithoutRepeats.Add(file, Path.GetFileName(file));
+            }
+        }
+        var classes = filesWithoutRepeats.Keys
+            .AsParallel()
+            .Select(Assembly.LoadFrom)
+            .SelectMany(a => a.ExportedTypes)
+            .Where(t => t.IsClass);
+        var classesWithTests =
+            classes.Where(t => t.GetMethods().Any(m => m.GetCustomAttributes().Any(a => a is Test)));
+        testsInfo = new ConcurrentQueue<TestInfo>();
+        Parallel.ForEach(classesWithTests, StartTests);
+    }
+
+    /// <summary> 
+    /// Starts tests for one class with test
+    /// </summary>
+    /// <param name="classWithTests">Class with test</param>
+    private void StartTests(Type classWithTests)
+    {
+        RunAfterClassOrBeforeClass(classWithTests, typeof(BeforeClass));
+        RunTests(classWithTests);
+        RunAfterClassOrBeforeClass(classWithTests, typeof(AfterClass));
+    }
+
+    /// <summary>
+    /// Run methods before or after all tests
+    /// </summary>
+    private void RunAfterClassOrBeforeClass(Type classWithTests, Type typeOfAttribute)
+    {
+        var methods = GetMethodsWithAttribute(classWithTests, typeOfAttribute);
+        foreach (var method in methods)
+        {
+            if (!method.IsStatic)
+            {
+                throw new InvalidOperationException();
+            }
+            try
+            {
+                method.Invoke(null, null);
+            }
+            catch (TargetInvocationException ex)
+            {
+                throw new AggregateException(ex);
```

Comment:
```suggestion
                throw new AggregateException(ex.InnerException);
```
Потому что суть ошибки у TargetInvocationException внутри